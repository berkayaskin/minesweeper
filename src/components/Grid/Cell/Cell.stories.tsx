import { Meta, Story } from '@storybook/react/types-6-0'
import { Cell, CellProps } from './Cell'

export default {
  title: 'Grid/Cell',
  component: Cell,
  argTypes: {
    coordinates: { defaultValue: { x: 1, y: 1 } },
  },
} as Meta

const Template: Story<CellProps> = (args) => <Cell {...args} />

export const Closed = Template.bind({})
Closed.args = {
  children: 10,
}

export const Empty = Template.bind({})
Empty.args = {
  children: 0,
}

export const Bomb = Template.bind({})
Bomb.args = {
  children: 9,
}

export const Flag = Template.bind({})
Flag.args = {
  children: 11,
}

export const WeakFlag = Template.bind({})
WeakFlag.args = {
  children: 12,
}

export const One = Template.bind({})
One.args = {
  children: 1,
}

export const Two = Template.bind({})
Two.args = {
  children: 2,
}

export const Three = Template.bind({})
Three.args = {
  children: 3,
}

export const Four = Template.bind({})
Four.args = {
  children: 4,
}

export const Five = Template.bind({})
Five.args = {
  children: 5,
}

export const Six = Template.bind({})
Six.args = {
  children: 6,
}

export const Seven = Template.bind({})
Seven.args = {
  children: 7,
}

export const Eight = Template.bind({})
Eight.args = {
  children: 8,
}
