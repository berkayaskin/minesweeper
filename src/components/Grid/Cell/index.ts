export { Cell, ClosedFrame, isActiveCell } from './Cell'
export type { CellProps } from './Cell'
