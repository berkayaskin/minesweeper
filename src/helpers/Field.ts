export type Cell = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12
export type Field = Cell[][]
export type Coords = [number, number]

export const CellState: Record<string, Cell> = {
  empty: 0,
  bomb: 9,
  hidden: 10,
  mark: 11,
  weakMark: 12,
}
