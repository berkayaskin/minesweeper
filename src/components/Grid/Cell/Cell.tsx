import { Cell as CellType, CellState, Coordinates } from '@/helpers/Field'
import { useMouseDown } from '@/hooks/useMouseDown'
import styled from '@emotion/styled'

export interface CellProps {
  /**
   * Cell status based on the CellType
   */
  children: CellType
  /**
   * Cell coordinates
   * @example [1, 2]
   */
  coordinates: Coordinates
  /**
   * onClick handler
   */
  onClick: (coordinates: Coordinates) => void
  /**
   * onContextMenu handler
   */
  onContextMenu: (coordinates: Coordinates) => void
}

export const isActiveCell = (cell: CellType): boolean =>
  [CellState.hidden, CellState.flag, CellState.weakFlag].includes(cell)

export const Cell = ({ children, coordinates, ...rest }: CellProps) => {
  const [mouseDown, onMouseDown, onMouseUp] = useMouseDown()

  const onClick = () => rest.onClick(coordinates)

  const onContextMenu = (e: React.MouseEvent<HTMLElement>) => {
    e.preventDefault()

    if (isActiveCell(children)) {
      rest.onContextMenu(coordinates)
    }
  }

  const props = {
    onClick,
    onContextMenu,
    onMouseDown,
    onMouseUp,
    onMouseLeave: onMouseUp,
    mouseDown,
    'data-testid': `${children}_${coordinates}`,
  }

  return <ComponentsMap {...props}>{children}</ComponentsMap>
}

interface ComponentsMapProps {
  children: CellType
  onClick: (elem: React.MouseEvent<HTMLElement>) => void
  onContextMenu: (elem: React.MouseEvent<HTMLElement>) => void
  onMouseDown: (elem: React.MouseEvent<HTMLElement>) => void
  onMouseUp: (elem: React.MouseEvent<HTMLElement>) => void
  onMouseLeave: (elem: React.MouseEvent<HTMLElement>) => void
  mouseDown: boolean
  'data-testid'?: string
}

const ComponentsMap = ({ children, ...rest }: ComponentsMapProps) => {
  const nonActiveCellProps = {
    onContextMenu: rest.onContextMenu,
    'data-testid': rest['data-testid'],
  }

  switch (children) {
    case CellState.empty:
      return <RevealedFrame {...nonActiveCellProps} />
    case CellState.bomb:
      return (
        <BombFrame {...nonActiveCellProps}>
          <Bomb />
        </BombFrame>
      )
    case CellState.hidden:
      return <ClosedFrame {...rest} />
    case CellState.flag:
      return (
        <ClosedFrame {...rest}>
          <Flag />
        </ClosedFrame>
      )
    case CellState.weakFlag:
      return (
        <ClosedFrame {...rest}>
          <WeakFlag />
        </ClosedFrame>
      )
    default:
      return <RevealedFrame {...nonActiveCellProps}>{children}</RevealedFrame>
  }
}

interface ClosedFrameProps {
  mouseDown?: boolean
}

export const ClosedFrame = styled.div<ClosedFrameProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 1.8vw;
  height: 1.8vw;
  cursor: pointer;
  user-select: none;
  background-color: #d1d1d1;
  border: 0.4vh solid transparent;
  border-color: ${({ mouseDown = false }) =>
    mouseDown ? '#transparent' : '#fff #9e9e9e #9e9e9e #fff'};
  &:hover {
    filter: brightness(1.1);
  }
`
const transparent = 'rgba(0, 0, 0, 0)'
const colors: { [key in CellType]: string } = {
  0: '#000000',
  1: '#2a48ec',
  2: '#2bb13d',
  3: '#ec6561',
  4: '#233db7',
  5: '#a6070f',
  6: '#e400af',
  7: '#906a02',
  8: '#fa0707',
  9: transparent,
  10: transparent,
  11: transparent,
  12: transparent,
}

const RevealedFrame = styled(ClosedFrame)`
  border-color: #dddddd;
  cursor: default;
  color: ${({ children }) => colors[children as CellType] ?? transparent};
  &:hover {
    filter: brightness(1);
  }
`
const Bomb = styled.div`
  width: 1vw;
  height: 1vw;
  background-color: #333;
  border-radius: 50%;
`

const BombFrame = styled(RevealedFrame)`
  background-color: #ec433c;
`

const Flag = styled.div`
  width: 0;
  height: 0;
  border-left: 0.5vw solid #ec433c;
  border-top: 0.5vw solid transparent;
  border-bottom: 0.5vw solid transparent;
`

const WeakFlag = styled(Flag)`
  border-left: 0.5vw solid #f19996;
`
