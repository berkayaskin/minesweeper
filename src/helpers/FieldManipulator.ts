import { CellState, Coordinates, Field } from './Field'

export const getNeighbours = ([x, y]: Coordinates): Record<string, [number, number]> => {
  return {
    topLeft: [x - 1, y - 1],
    top: [x - 1, y],
    topRight: [x - 1, y + 1],
    left: [x, y - 1],
    right: [x, y + 1],
    bottomLeft: [x + 1, y - 1],
    bottom: [x + 1, y],
    bottomRight: [x + 1, y + 1],
  }
}

export const isItemExistInField = ({ length }: Field, [x, y]: Coordinates): boolean => {
  return x >= 0 && x < length && y >= 0 && y < length
}

export const incrementNeighbours = (field: Field, coordinates: Coordinates): Field => {
  const neighbours = getNeighbours(coordinates)
  const newField = [...field]

  Object.values(neighbours).forEach(([x, y]) => {
    if (isItemExistInField(field, [x, y])) {
      if (newField[x][y] !== CellState.bomb) {
        newField[x][y]++
      }
    }
  })

  return newField
}
