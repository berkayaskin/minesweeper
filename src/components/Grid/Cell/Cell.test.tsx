import { CellState, Coordinates } from '@/helpers/Field'
import { createEvent, fireEvent, render, screen } from '@testing-library/react'
import { Cell, ClosedFrame, isActiveCell } from './Cell'

describe('Cell', () => {
  const coordinates: Coordinates = [1, 1]

  const props = {
    coordinates,
    onClick: jest.fn(),
    onContextMenu: jest.fn(),
  }

  for (let cell = CellState.empty; cell <= CellState.weakFlag; cell++) {
    it('should render cell correctly', () => {
      const { asFragment } = render(<Cell {...props}>{cell}</Cell>)
      expect(asFragment()).toMatchSnapshot()
    })

    it('should render closed frame correctly', () => {
      const { asFragment } = render(<ClosedFrame mouseDown={true} />)
      expect(asFragment()).toMatchSnapshot()
    })

    it('should prevent default context menu for all type of cell', () => {
      render(<Cell {...props}>{cell}</Cell>)

      const cellComponent = screen.getByTestId(`${cell}_${coordinates}`)
      const contextMenuEvent = createEvent.contextMenu(cellComponent)
      fireEvent(cellComponent, contextMenuEvent)

      expect(contextMenuEvent.defaultPrevented).toBe(true)
    })

    it('should call onClick and onContextMenu for active cell', () => {
      render(<Cell {...props}>{cell}</Cell>)

      const cellComponent = screen.getByTestId(`${cell}_${coordinates}`)
      fireEvent.click(cellComponent)
      fireEvent.contextMenu(cellComponent)

      if (isActiveCell(cell)) {
        expect(props.onClick).toBeCalledWith(coordinates)
        expect(props.onContextMenu).toBeCalledWith(coordinates)
      } else {
        expect(props.onClick).not.toBeCalled()
        expect(props.onContextMenu).not.toBeCalled()
      }
    })
  }
})
