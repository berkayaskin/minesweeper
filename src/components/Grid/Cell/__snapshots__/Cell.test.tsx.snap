// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell should render cell correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1bascby"
    data-testid="0_1,1"
  />
</DocumentFragment>
`;

exports[`Cell should render cell correctly 2`] = `
<DocumentFragment>
  <div
    class="css-1yu2ia5"
    data-testid="1_1,1"
  >
    1
  </div>
</DocumentFragment>
`;

exports[`Cell should render cell correctly 3`] = `
<DocumentFragment>
  <div
    class="css-1w4pte2"
    data-testid="2_1,1"
  >
    2
  </div>
</DocumentFragment>
`;

exports[`Cell should render cell correctly 4`] = `
<DocumentFragment>
  <div
    class="css-1e2a6pr"
    data-testid="3_1,1"
  >
    3
  </div>
</DocumentFragment>
`;

exports[`Cell should render cell correctly 5`] = `
<DocumentFragment>
  <div
    class="css-1qy3cpa"
    data-testid="4_1,1"
  >
    4
  </div>
</DocumentFragment>
`;

exports[`Cell should render cell correctly 6`] = `
<DocumentFragment>
  <div
    class="css-596k7x"
    data-testid="5_1,1"
  >
    5
  </div>
</DocumentFragment>
`;

exports[`Cell should render cell correctly 7`] = `
<DocumentFragment>
  <div
    class="css-1kwfzmq"
    data-testid="6_1,1"
  >
    6
  </div>
</DocumentFragment>
`;

exports[`Cell should render cell correctly 8`] = `
<DocumentFragment>
  <div
    class="css-dx6f97"
    data-testid="7_1,1"
  >
    7
  </div>
</DocumentFragment>
`;

exports[`Cell should render cell correctly 9`] = `
<DocumentFragment>
  <div
    class="css-mjygpe"
    data-testid="8_1,1"
  >
    8
  </div>
</DocumentFragment>
`;

exports[`Cell should render cell correctly 10`] = `
<DocumentFragment>
  <div
    class="css-1na1t6s"
    data-testid="9_1,1"
  >
    <div
      class="css-b14p9c"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell should render cell correctly 11`] = `
<DocumentFragment>
  <div
    class="css-rqpdce"
    data-testid="10_1,1"
  />
</DocumentFragment>
`;

exports[`Cell should render cell correctly 12`] = `
<DocumentFragment>
  <div
    class="css-rqpdce"
    data-testid="11_1,1"
  >
    <div
      class="css-3cmri2"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell should render cell correctly 13`] = `
<DocumentFragment>
  <div
    class="css-rqpdce"
    data-testid="12_1,1"
  >
    <div
      class="css-vtfnhb"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell should render closed frame correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1wtra4k"
  />
</DocumentFragment>
`;

exports[`Cell should render closed frame correctly 2`] = `
<DocumentFragment>
  <div
    class="css-1wtra4k"
  />
</DocumentFragment>
`;

exports[`Cell should render closed frame correctly 3`] = `
<DocumentFragment>
  <div
    class="css-1wtra4k"
  />
</DocumentFragment>
`;

exports[`Cell should render closed frame correctly 4`] = `
<DocumentFragment>
  <div
    class="css-1wtra4k"
  />
</DocumentFragment>
`;

exports[`Cell should render closed frame correctly 5`] = `
<DocumentFragment>
  <div
    class="css-1wtra4k"
  />
</DocumentFragment>
`;

exports[`Cell should render closed frame correctly 6`] = `
<DocumentFragment>
  <div
    class="css-1wtra4k"
  />
</DocumentFragment>
`;

exports[`Cell should render closed frame correctly 7`] = `
<DocumentFragment>
  <div
    class="css-1wtra4k"
  />
</DocumentFragment>
`;

exports[`Cell should render closed frame correctly 8`] = `
<DocumentFragment>
  <div
    class="css-1wtra4k"
  />
</DocumentFragment>
`;

exports[`Cell should render closed frame correctly 9`] = `
<DocumentFragment>
  <div
    class="css-1wtra4k"
  />
</DocumentFragment>
`;

exports[`Cell should render closed frame correctly 10`] = `
<DocumentFragment>
  <div
    class="css-1wtra4k"
  />
</DocumentFragment>
`;

exports[`Cell should render closed frame correctly 11`] = `
<DocumentFragment>
  <div
    class="css-1wtra4k"
  />
</DocumentFragment>
`;

exports[`Cell should render closed frame correctly 12`] = `
<DocumentFragment>
  <div
    class="css-1wtra4k"
  />
</DocumentFragment>
`;

exports[`Cell should render closed frame correctly 13`] = `
<DocumentFragment>
  <div
    class="css-1wtra4k"
  />
</DocumentFragment>
`;
